File: src/tree/Node.java
package tree;

// todo parameterize class Node
public class Node {

    private final int    key;
    private final Object val;

    Node(Builder builder)
    {
        key = builder.key;
        val = builder.val;
    }

    // getter
    public int getKey()
    {
        return key;
    }

    public Object getVal()
    {
        return val;
    }
    // getter end

    @Override
    public String toString()
    {
        return "Node{" + "key=" + key + ", val=" + val + '}';
    }

    public static class Builder {

        private final int    key;
        private final Object val;

        public Builder(int key, Object val)
        {
            this.key = key;
            this.val = val;
        }

        public Node build()
        {
            return new Node(this);
        }

    }

}

File: src/tree/Node.kt
package tree

// suggest parameterize class Node
open class Node internal constructor(
    builder: Builder,
) {
    val key: Int = builder.key
    val value: Any? = builder.value

    override fun toString(): String = "Node{key=$key, val=$value}"

    open class Builder(
        internal val key: Int,
        val value: Any?,
    ) {
        open fun build(): Node? = Node(this)
    }
}
